src/libcharon/plugins/kernel_super/system_events/sl_system_events/sl_system_events.c

```diff
@@ -0,0 +1,153 @@
+#include "sl_system_events.h"
+
+#include "../libsystem_events/libsystem_events.h"
+
+#include <dlfcn.h>
+#include <stdlib.h>
+#include <string.h>
+#include <utils/debug.h>
+
+#define LIBSYSTEMEVENTS_SO_PATH "/usr/local/lib/libsystem_events.so"
+
+typedef struct private_sl_system_events_t private_sl_system_events_t;
+
+struct private_sl_system_events_t {
+	void *dl_handle;
+	sl_system_events_obj_t public;
+	struct {
+		typeof(libsystem_events_send_ipsec_tunnel_established)
+			*libsystem_events_send_ipsec_tunnel_established;
+		typeof(libsystem_events_send_ipsec_tunnel_establishment_failed)
+			*libsystem_events_send_ipsec_tunnel_establishment_failed;
+		typeof(libsystem_events_send_ipsec_tunnel_activated)
+			*libsystem_events_send_ipsec_tunnel_activated;
+		typeof(libsystem_events_send_ipsec_tunnel_deactivated)
+			*libsystem_events_send_ipsec_tunnel_deactivated;
+		typeof(libsystem_events_send_ipsec_tunnel_down) *libsystem_events_send_ipsec_tunnel_down;
+		typeof(
+			libsystem_events_send_ipsec_terminator_up) *libsystem_events_send_ipsec_terminator_up;
+		typeof(libsystem_events_send_ipsec_terminator_down)
+			*libsystem_events_send_ipsec_terminator_down;
+	} ops;
+};
+
+static private_sl_system_events_t *g_system_events = NULL;
+
+static int sl_send_ipsec_tunnel_established(const char *device_id,
+											const char *terminator_id,
+											const char *vrf_name)
+{
+	return g_system_events->ops.libsystem_events_send_ipsec_tunnel_established(device_id,
```

Comment:
Note all the functions in `system_event_t` are assumed to return 1 on success and 0 on failure (following `empty_system_events_t`), however the implementation of `libsystem_events` returns 0 on success and -1 on failure.

Don't use negative return values here. Instead make all the APIs return a boolean and make the necessary conversions in the `libsystem_events`